bip-0340.mediawiki

```diff
@@ -237,6 +237,25 @@ the algorithms of the signature scheme will perform more hashing internally.
 In particular, the hashing performed by the signing algorithm will process the message twice,
 which leads to performance penalties for large messages.
 
+==== Domain Separation ====
+
+It is good cryptographic practice to use a key pair only for a single purpose.
+Nevertheless, there may be situations in which it may be desirable to use the same key pair in multiple contexts,
+i.e., to sign different types of messages within the same application
+or even messages in entirely different applications
+(e.g., a secret key may be used to sign Bitcoin transactions as well plain text messages).
+
+As a consequence, applications should ensure that a signed application message intended for one context is never deemed valid in a different context
+(e.g., a signed plain text message should never be misinterpreted as a signed Bitcoin transaction, because this could cause unintended loss of funds).
+This is called "domain separation" and it is typically realized by partitioning the message space.
+Even if key pairs are intended to be used only within a single context,
+domain separation is a good idea because it makes it easy to add more contexts later.
+
+As a best practice, we recommend to pre-process application messages before passing it the signature scheme as follows:
+ - Either, pre-hash the application message using ''hash<sub>tag</sub>'', where the ''tag'' is a byte string that identifies the context uniquely (e.g., "my-fancy-application/signing-foo-messages"),
+ - or prefix the actual message with a 64-byte string that identifies the context uniquely (e.g. "my-fancy-application/signing-foo-messages", padded with null bytes to 64 bytes).
```

Comment:
why 64 bytes?

Reply:
This is @sipa's suggestion in https://github.com/sipa/bips/issues/207#issuecomment-673681901 and the outcome of the discussion.

I think the technical reason is that it's greater than 32 (no collisions with the other method) and a multiple of 32 (you can precompute the midstate).

Reply:
Precomputing the midstate only makes sense when you split up signing into a two-stage process where you first generate the nonce and create the challenge hash up to the tag. Second, you receive the message and use the midstate to finish the challenge hash computation quickly. This is prone to nonce reuse. Also, mandating a 64-byte is slower in total than 33-byte tags with short messages (<= 64-33-8 = 23 bytes) because an additional compression round is required. Signing short messages feels slightly more relevant than two-step signing.

In any case, I'd suggest adding a footnote that explains the choice. The 33-byte variant appears to have an easier explanation. As @real-or-random discussed elsewhere, an alternative is to suggest creating the 64-byte tag as `hash(tag) || hash(tag)`. Then it's obvious why it's 64 bytes, requires no padding, and allows using both 33-byte and 64-byte tags without the risk of collision. However, this adds the requirement that implementations call a hash function, reducing the appeal of the variable-length proposal.

Reply:
Updated to the 33-byte variant.

> Signing short messages feels slightly more relevant than two-step signing.

Indeed.  And if someone really needs "offline/online signatures" (that's the literature term for "two-step", where the first step doesn't have the message), they can still pad more.

> In any case, I'd suggest adding a footnote that explains the choice. The 33-byte variant appears to have an easier explanation.

I tried but then I felt the sentence "As the two pre-processing methods yield different message sizes (32 bytes vs. at least 33 bytes), there is no risk of collision between them." already hints so strongly at the explanation (namely that 33 is simply the smallest integer with that property) that it's not necessary to say this.


> As @real-or-random discussed elsewhere, an alternative is to suggest creating the 64-byte tag as `hash(tag) || hash(tag)`. Then it's obvious why it's 64 bytes, requires no padding, and allows using both 33-byte and 64-byte tags without the risk of collision. However, this adds the requirement that implementations call a hash function, reducing the appeal of the variable-length proposal.

Yes, I rejected that idea. Here are some more reasons. It could also raise questions about collision-resistance of `hash` again. And also, it has another strange risk of collision: It's not totally absurd that `hash(tag||m)` (first method) could collide with `hash(tag')||hash(tag')||m'` (second method) if `tag` is a prefix of `tag'`.

Reply:
> that it's not necessary to say this.

Yes, sorry, I misspoke. In this case a footnote is not necessary.

> it has another strange risk of collision

As far as I can see, the first method is actually `tag||m` (where tag is 33 bytes) so I don't see how they can collide.

Reply:
> As far as I can see, the first method is actually `tag||m` (where tag is 33 bytes) so I don't see how they can collide.

Hm lol, I think we are both wrong. 

> an alternative is to suggest creating the 64-byte tag as `hash(tag) || hash(tag)`

So if we took that approach,then the methods would, be:
1. `hash_tag(m) = hash(hash(tag) || hash(tag) || m)`
2. `hash(tag') || hash(tag') || m'`

But yeah, there are no way these could collide. So I take back that argument. The other arguments are still valid though

Reply:
Ok, makes sense. But just to be clear at least in the paragraph you're quoting I was in fact referring to the `tag||m` variant where `tag` is 33 bytes.